### PR TITLE
LASSCF 3-RDMs

### PR DESCRIPTION
**Pull request by shreya-chem**
Added a few functions for casdm2s and casdm3 in mcscf/lasci.py, plus a unit test for 3-RDMs on H8 with different sub-spins.

### CODE REVIEW

File: my_pyscf/mcscf/las_ao2mo.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+libmcscf = lib.load_library('libmcscf')
+
+def _mem_usage(las, mo_coeff, ncore, ncas, nmo):
+    ncore = las.ncore
+    ncas = las.ncas
+    nmo = mo_coeff.shape[-1]
+    nocc = ncore + ncas
+    outcore = basic = ncas**2*nmo**2*2 * 8/1e6
+    incore = outcore + (ncore+ncas)*nmo**3*4/1e6
+    return incore, outcore, basic
```
Comment (MatthewRHermes) [inline my_pyscf/mcscf/las_ao2mo.py:21]: This conflicts with my las_ao2mo.py. The conflict has to be resolved. I don't see why you need any of this for this PR.

File: my_pyscf/mcscf/lasci.py [hunk 2]
```diff
@@ -271,6 +272,7 @@ def h1e_for_las (las, mo_coeff=None, ncas=None, ncore=None, nelecas=None, ci=Non
     h1e_r = np.empty ((las.nroots, 2, ncas, ncas), dtype=h1e.dtype)
     h2e = lib.numpy_helper.unpack_tril (h2eff_sub.reshape (nmo*ncas,
         ncas*(ncas+1)//2)).reshape (nmo, ncas, ncas, ncas)[ncore:nocc,:,:,:]
+    #print ("SV h2e = ",h2e, h2e.shape)
```
Comment (MatthewRHermes) [inline my_pyscf/mcscf/lasci.py:275]: Plz delete this

File: my_pyscf/mcscf/lasci.py [hunk 3]
```diff
@@ -1043,6 +1045,101 @@ def make_casdm2_sub (self, ci=None, ncas_sub=None, nelecas_sub=None, casdm2fr=No
...
+        for fcibox, ci_i, ncas, nelecas in zip (self.fciboxes, ci, ncas_sub, nelecas_sub):
+            if ci_i is None:
+                dm2aa = dm2ab, dm2bb = np.zeros ((ncas, ncas,ncas,ncas))
+            else:
+                dm2aa, dm2ab, dm2bb = fcibox.states_make_rdm12s (ci_i, ncas, nelecas)[-1]
+            casdm2s.append (np.stack ([dm2aa, dm2ab, dm2bb], axis=1))
+        return casdm2s
+    
+    #SV states_make_casdm2s
+    def states_make_casdm2s (self, ci=None, ncas_sub=None, nelecas_sub=None,
```
Comment (MatthewRHermes) [inline my_pyscf/mcscf/lasci.py:1076]: Why is this function necessary?

File: my_pyscf/mcscf/lasci.py [hunk 4]
```diff
@@ -1043,6 +1045,101 @@ def make_casdm2_sub (self, ci=None, ncas_sub=None, nelecas_sub=None, casdm2fr=No
...
+                    casdm2[k:l, k:l, i:j, i:j] = casdm2[i:j, i:j, k:l, k:l].transpose (2,3,0,1)
+                    # Exchange slice
+                    casdm2[i:j, k:l, k:l, i:j] = -(np.multiply.outer (dma1, dma2)
+                                               +np.multiply.outer (dmb1, dmb2)).transpose (0,3,2,1)
+                    casdm2[k:l, i:j, i:j, k:l] = casdm2[i:j, k:l, k:l, i:j].transpose (1,0,3,2)
+
+        return casdm2rs
+    
+    #SV make_casdm3_sub
+    def make_casdm3_sub (self, ci=None, ncas_sub=None, nelecas_sub=None, casdm3fr=None, **kwargs):
```
Comment (MatthewRHermes) [inline my_pyscf/mcscf/lasci.py:1119]: Why is this function necessary?

**Review by MatthewRHermes — CHANGES_REQUESTED** [review]
Please remove all "print" lines and all commented-out code (not genuine comments, just code that you commented-out)

File: my_pyscf/mcscf/lasci.py (hunk 4 above)
Comment (shreya-chem) [reply my_pyscf/mcscf/lasci.py]: I didn't use it in make_casdm3 but while understanding the workflow, I just made it. It is not needed. I have removed it.

**Comment by MatthewRHermes** [comment]
Looks good, merging this. Thanks!

**Pull request merged**